plugins/org.jkiss.dbeaver.ui.editors.sql/src/org/jkiss/dbeaver/ui/editors/sql/generator/SQLGeneratorDialog.java

```diff
@@ -99,11 +110,48 @@ protected Composite createDialogArea(Composite parent) {
         sqlGenerator.setShowFullDdl(getDialogBoundsSettings().get(DBPScriptObject.OPTION_INCLUDE_NESTED_OBJECTS) != null &&
                 getDialogBoundsSettings().getBoolean(DBPScriptObject.OPTION_INCLUDE_NESTED_OBJECTS));
 
-        UIUtils.runInUI(sqlGenerator);
-        Object sql = sqlGenerator.getResult();
-        if (sql != null) {
-            setSQLText(CommonUtils.toString(sql));
-        }
+        generateDDLJob = new AbstractJob("Generating DDL") {
+            @Override
+            protected IStatus run(DBRProgressMonitor monitor) {
+                try {
+                    if (SQLGeneratorDialog.this.getShell() != null && !SQLGeneratorDialog.this.getShell().isDisposed()) {
+                        UIUtils.syncExec(() -> {
+                            setSQLText("Loading DDL...");
+                            updateSQL();
+                            Button button = getButton(IDialogConstants.DETAILS_ID);
+                            if (button != null && !button.isDisposed()) {
+                                button.setEnabled(false);
+                            }
+                        });
+                    }
+                    sqlGenerator.run(monitor);
+                    Object sql = sqlGenerator.getResult();
+                    if (
```

Comment:
Move all checks inside syncExec (never operate with shell outside of UI thread. Even getShell function may fail).
No need to check isDisposed for button. Use button variable from block above instead of getButton.